extractor/image/image.go

```diff
@@ -0,0 +1,240 @@
+		var rawSource imageTypes.ImageSource
+		rawSource, err = ref.NewImageSource(ctx, img.systemContext)
+		if err != nil {
+			// try next transport
+			continue
+		}
+
+		var src imageTypes.ImageCloser
+		src, err = image.FromSource(ctx, img.systemContext, rawSource)
+		if err != nil {
+			return xerrors.Errorf("failed to initialize: %w", err)
+		}
+
+		img.rawSource = rawSource
+		img.src = src
+		return nil
+	}
+	// return only the last error
+	return err
+}
+
+func (img *Image) LayerInfos() ([]imageTypes.BlobInfo, error) {
+	// ignore the cache error
+	layerInfos, _ := img.getLayerInfosInCache(img.name, img.isFile)
+	if layerInfos != nil {
+		return layerInfos, nil
+	}
+
+	// When it doesn't hit a cache, it fetches the image from Docker Engine or Docker Registry
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	layers := img.src.LayerInfos()
+
+	// ignore the cache error
+	_ = img.storeLayerInfosInCache(img.name, img.isFile, layers)
+
+	return layers, nil
+}
+
+func (img Image) getLayerInfosInCache(imageName string, isFile bool) ([]imageTypes.BlobInfo, error) {
+	// tar file doesn't have info in cache
+	if isFile {
+		return nil, nil
+	}
+	rc := img.cache.Get("layerinfos::" + imageName)
+	if rc == nil {
+		return nil, nil
+	}
+	var layerInfos []imageTypes.BlobInfo
+	if err := json.NewDecoder(rc).Decode(&layerInfos); err != nil {
+		return nil, xerrors.Errorf("invalid cache: %w", err)
+	}
+	return layerInfos, nil
+}
+
+func (img Image) storeLayerInfosInCache(imageName string, isFile bool, layers []imageTypes.BlobInfo) error {
+	// it doesn't store the information of a tar file
+	if isFile {
+		return nil
+	}
+
+	b, err := json.Marshal(layers)
+	if err != nil {
+		return xerrors.Errorf("invalid json: %w", err)
+	}
+	return img.cache.SetBytes("layerinfos::"+imageName, b)
+}
+
+func (img *Image) ConfigBlob(ctx context.Context) ([]byte, error) {
+	b, _ := img.getConfigBlobInCache(img.name, img.isFile)
+	if b != nil {
+		return b, nil
+	}
+
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	blob, err := img.src.ConfigBlob(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get a config: %w", err)
+	}
+
+	if !img.isFile {
+		_ = img.cache.SetBytes("configblob::"+img.name, blob)
+	}
+
+	return blob, nil
+}
+
+func (img Image) getConfigBlobInCache(imageName string, isFile bool) ([]byte, error) {
+	// tar file doesn't have info in cache
+	if isFile {
+		return nil, nil
+	}
+	rc := img.cache.Get("configblob::" + imageName)
+	if rc == nil {
+		return nil, nil
+	}
+	b, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return nil, xerrors.Errorf("unable to read a cache file: %w", err)
+	}
+	return b, nil
+}
+
+func (img *Image) GetBlob(ctx context.Context, dig digest.Digest) (io.ReadCloser, error) {
+	rc := img.cache.Get(dig.String())
+	if rc != nil {
+		return rc, nil
+	}
+
+	if err := img.populateSource(); err != nil {
```

Comment:
`populateSource()` doesn't seem to be unit tested on it's own nor does it seem to have a mock in a situation like this call site. We should add both as it seems to be an important function that's commonly used by various other functions.

Reply:
I fixed each test not to skip `populateSource`. I feel we can test this function enough via other unit tests.

Reply:
Looks much better now – thanks!